keanu-project/src/main/java/io/improbable/keanu/algorithms/graphtraversal/BreadthFirstSearch.java

```diff
@@ -0,0 +1,52 @@
+package io.improbable.keanu.algorithms.graphtraversal;
+
+import io.improbable.keanu.vertices.Vertex;
+
+import java.util.ArrayDeque;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Queue;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+
+public class BreadthFirstSearch {
+
+    public static boolean bfs(Collection<Vertex> vertices,
+                              Predicate<Vertex> failureCondition,
+                              Predicate<Vertex> shouldAddParents,
```

Comment:
You can traverse in either direction. It might not always be right to call it parents, it could be children.

Reply:
Good point. I will replace this with a `Function<Vertex, Collection<Vertex>>` that will calculate the next vertices to visit. That way the predicate `shouldAddParents` can disappear and instead be represented in the function by returning an empty collection when parents don't need to be added.